pkg/graveler/graveler.go

```diff
@@ -1606,29 +1629,50 @@ func (g *KVGraveler) List(ctx context.Context, repositoryID RepositoryID, ref Re
 }
 
 func (g *KVGraveler) Commit(ctx context.Context, repositoryID RepositoryID, branchID BranchID, params CommitParams) (CommitID, error) {
-	// TODO(issue 3569) - kv implementation optimistic lock commit
 	var preRunID string
 	var commit Commit
+	var newCommitID CommitID
 	var storageNamespace StorageNamespace
-	res, err := g.db.branchLocker.MetadataUpdater(ctx, repositoryID, branchID, func() (interface{}, error) {
-		isProtected, err := g.protectedBranchesManager.IsBlocked(ctx, repositoryID, branchID, BranchProtectionBlockedAction_COMMIT)
+
+	isProtected, err := g.protectedBranchesManager.IsBlocked(ctx, repositoryID, branchID, BranchProtectionBlockedAction_COMMIT)
+	if err != nil {
+		return "", err
+	}
+	if isProtected {
+		return "", ErrCommitToProtectedBranch
+	}
+	repo, err := g.RefManager.GetRepository(ctx, repositoryID)
+	if err != nil {
+		return "", fmt.Errorf("get repository: %w", err)
+	}
+	storageNamespace = repo.StorageNamespace
+
+	branch, err := g.RefManager.GetBranch(ctx, repositoryID, branchID)
+	if err != nil {
+		return "", fmt.Errorf("get branch: %w", err)
+	}
+
+	if params.SourceMetaRange != nil {
```

Comment:
I think this check should be part of the branch update as we like to verify that the branch state doesn't change during this one or verify that the branch information read as part of UpdateBranch wasn't modified as we checked empty staging on specific tokens.

Reply:
done

Reply:
It was because of another comment of Itai to check it before the branchUpdate as well, but we decided to delete it. It is checked in the branchUpdate too.